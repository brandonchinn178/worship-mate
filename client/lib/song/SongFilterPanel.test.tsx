import { fireEvent, waitFor } from '@testing-library/react'

import { renderUI } from '~jest-utils'

import { SongFilterPanel } from './SongFilterPanel'

const testFilters = {
  recommendedKey: [
    {
      value: 'C',
      valueDisplay: 'C',
      count: 1,
    },
    {
      value: 'G',
      valueDisplay: 'G',
      count: 2,
    },
    {
      value: 'E',
      valueDisplay: 'E',
      count: 3,
    },
  ],
  bpm: [
    {
      value: 100,
      valueDisplay: '100',
      count: 6,
    },
  ],
  timeSignature: [
    {
      value: [4, 4] as [number, number],
      valueDisplay: '4/4',
      count: 6,
    },
  ],
}

const mockFilterHandler = {
  addFilter: jest.fn(),
  removeFilter: jest.fn(),
}

beforeEach(jest.resetAllMocks)

it('renders filters', () => {
  const { getByText } = renderUI(
    <SongFilterPanel
      availableFilters={testFilters}
      activeFilters={{}}
      filterHandler={mockFilterHandler}
    />,
  )

  expect(getByText('Recommended Key')).toBeVisible()
  expect(getByText('C (1)')).toBeVisible()
  expect(getByText('G (2)')).toBeVisible()
  expect(getByText('E (3)')).toBeVisible()
  expect(getByText('Time Signature')).toBeVisible()
  expect(getByText('4/4 (6)')).toBeVisible()
})

it('can add a filter', async () => {
  const { getByText } = renderUI(
    <SongFilterPanel
      availableFilters={testFilters}
      activeFilters={{}}
      filterHandler={mockFilterHandler}
    />,
  )

  fireEvent.click(getByText('G (2)'))

  const mockAddFilter = mockFilterHandler.addFilter
  await waitFor(() => expect(mockAddFilter).toHaveBeenCalled())
  expect(mockAddFilter.mock.calls[0]).toEqual(['recommendedKey', 'G'])
})

it('can change a filter', async () => {
  const { getByText } = renderUI(
    <SongFilterPanel
      availableFilters={testFilters}
      activeFilters={{ recommendedKey: 'G' }}
      filterHandler={mockFilterHandler}
    />,
  )

  fireEvent.click(getByText('E (3)'))

  const mockAddFilter = mockFilterHandler.addFilter
  await waitFor(() => expect(mockAddFilter).toHaveBeenCalled())
  expect(mockAddFilter.mock.calls[0]).toEqual(['recommendedKey', 'E'])
})

it('can remove a filter', async () => {
  const { getByText } = renderUI(
    <SongFilterPanel
      availableFilters={testFilters}
      activeFilters={{ recommendedKey: 'G' }}
      filterHandler={mockFilterHandler}
    />,
  )

  fireEvent.click(getByText('G (2)'))

  const mockRemoveFilter = mockFilterHandler.removeFilter
  await waitFor(() => expect(mockRemoveFilter).toHaveBeenCalled())
  expect(mockRemoveFilter.mock.calls[0]).toEqual(['recommendedKey'])
})

it('renders active options differently', () => {
  const { container: before } = renderUI(
    <SongFilterPanel
      availableFilters={testFilters}
      activeFilters={{}}
      filterHandler={mockFilterHandler}
    />,
  )

  const { container: after } = renderUI(
    <SongFilterPanel
      availableFilters={testFilters}
      activeFilters={{ recommendedKey: 'G' }}
      filterHandler={mockFilterHandler}
    />,
  )

  expect(before).toMatchDiffSnapshot(after)
})
