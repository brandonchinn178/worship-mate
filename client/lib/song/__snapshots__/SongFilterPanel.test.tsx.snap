// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders active options differently 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -20,11 +20,11 @@
             (
            1
            )
          </div>
          <div
-           class=\\"SongFilterPanel__FilterOption-sc-1esthvi-4 kafzeo\\"
+           class=\\"SongFilterPanel__FilterOption-sc-1esthvi-4 iYSQTL\\"
          >
            G
             (
            2
            )"
`;
